File: genwavfm.rb
#!/usr/bin/env ruby
require 'mp3info'
require 'optparse'

# Parse arguments
options = {}
optparse = OptionParser.new do |opts|
  opts.banner = 'Usage: genwavfm.rb --input PATH --output PATH'

  opts.on('-i', '--input PATH', 'Input path') do |i|
    options[:input] = i
  end

  opts.on('-o', '--output PATH', 'Output path') do |o|
    options[:output] = o
  end
end
optparse.parse!

# Ensure directory paths end in slashes
options[:input] += '/' unless options[:input][-1] == '/'
options[:output] += '/' unless options[:output][-1] == '/'

mp3s = Dir.glob(options[:input] + '**/*.mp3')
mp3s.each do |f|
  # Get the artist and album from the filepath
  artist = File.basename(File.dirname(File.dirname(f)))
  album = File.basename(File.dirname(f))

  # Workaround for audiowaveform bug; seemingly it should read length
  # correctly and automatically, but it doesn't for me
  length = Mp3Info.open(f).length
  length = length.ceil

  if artist == 'Various Artists' # Soundtracks and compilations
    output_filename = 'VA_' + album + '_' + File.basename(f)
    output_filename.gsub!(/ /, '_')
    output_filename.gsub!(/mp3/, 'png')
  else # All else
    output_filename = artist + '_' + album + '_' + File.basename(f)
    output_filename.gsub!(/ /, '_')
    output_filename.gsub!(/mp3/, 'png')
  end

  output_filename = options[:output] + output_filename

  system('audiowaveform', '-i', File.absolute_path(f), '-o', output_filename, '-e', length.to_s, '-w', '1600', '-h', '500')
end

File: genwavfm.nim
#!/usr/bin/env nim
import math, os, ospaths, osproc, parseopt, re, strutils

# Parse arguments
var options: tuple[inputPath: string, outputPath: string]
for kind, key, val in getopt():
  case kind
  of cmdArgument:
    continue
  of cmdLongOption, cmdShortOption:
    case key
    of "input", "i": options.inputPath = val
    of "output", "o": options.outputPath = val
  of cmdEnd: assert(false)

# Ensure directory paths end in slashes
if options.inputPath[^1] != '/':
  options.inputPath = options.inputPath & "/"
if options.outputPath[^1] != '/':
  options.outputPath = options.outputPath & "/"

# Main part of program
for file in walkDirRec(options.inputPath):
  if file.contains(re"mp3") == true:
    var spPath = file.split(re"/")
    var artist = spPath[4]
    var album = spPath[5]

    var result = execProcess("ffprobe -v error -show_entries format=duration \"$1\"" % [file])
    var lmArr = result.findAll(re"\d+\.\d+")
    var length = ceil(parseFloat(lmArr[0]))
    var outputFile: string

    if artist == "Various Artists":
      outputFile = "VA_" & album & "_" & spPath[6]
    else:
      outputFile = artist & "_" & album & "_" & spPath[6]

    outputFile = outputFile.replace(" ", "_")
    outputFile = outputFile.replace("mp3", "png")
    outputFile = options.outputPath & outputFile

    let errC = execCmd("audiowaveform -i \"$1\" -o \"$2\" -e $3 -w 1600 -h 500" % [file, outputFile, length.formatFloat(ffDecimal, 3)])
